tests/pytests/test_empty.py

```diff
@@ -888,3 +681,625 @@ def testEmptyInfo():
     env.assertEqual(tag_info[-1], 'INDEXEMPTY')
     text_info = info['attributes'][1]
     env.assertEqual(text_info[-1], 'INDEXEMPTY')
+
+def testEmptyNotActivated():
+    """Tests that the indexing of empty values is not activated by default, and
+    that no results are returned when searching for empty values in case the
+    feature is not activated."""
+
+    env = Env(moduleArgs="DEFAULT_DIALECT 2")
+
+    # Create an index with the currently supported field types (TAG, TEXT)
+    env.expect('FT.CREATE', 'idx', 'SCHEMA', 't', 'TAG', 'text', 'TEXT').ok()
+
+    # Populate the db with a document that has an empty value for a TAG field
+    conn = getConnectionByEnv(env)
+    conn.execute_command('HSET', 'h1', 't', '')
+
+    # Search for the document, it shouldn't be found
+    res = conn.execute_command('FT.SEARCH', 'idx', '@t:{""}')
+    expected = EMPTY_RESULT
+    env.assertEqual(res, expected)
+
+    # Search for the document, it shouldn't be found
+    res = conn.execute_command('FT.SEARCH', 'idx', '@t:""')
+    env.assertEqual(res, expected)
+
+def testEmptyExplainCli(env):
+    """Tests the output of `FT.EXPAINCLI` for queries that include empty values,
+    for both TAG and TEXT fields, for all supporting dialects ({2 ,3, 4, 5})."""
+
+    MAX_DIALECT = set_max_dialect(env)
+    for dialect in range(2, MAX_DIALECT + 1):
+        env = Env(moduleArgs="DEFAULT_DIALECT " + str(dialect))
```

Comment:
Overriding `env` is bad practice, and also in the case of failure, it's hard to know which dialect has failed.
Consider reimplementing the test so each dialect gets its own variant:
```python
def doEmptyExplainCli(dialect):
    env = Env(moduleArgs="DEFAULT_DIALECT " + str(dialect))
    # < single test iteration >

# Returns a lambda that gets no parameters and calls the main test with the right dialect
EmptyExplainCli_gen = lambda dialect: lambda: doEmptyExplainCli(dialect)
for dialect in range(2, MAX_DIALECT + 1):
    globals()[f'testEmptyExplainCli_Dialect_{dialect}'] = EmptyExplainCli_gen(dialect)
```
I suggest applying this to all the new tests

Reply:
Alternatively, you can change the configuration instead
```suggestion
def testEmptyExplainCli(env):
    """Tests the output of `FT.EXPAINCLI` for queries that include empty values,
    for both TAG and TEXT fields, for all supporting dialects ({2 ,3, 4, 5})."""

    MAX_DIALECT = set_max_dialect(env)
    for dialect in range(2, MAX_DIALECT + 1):
        env.expect(config_cmd(), 'SET', 'DEFAULT_DIALECT', dialect).ok()
```

Reply:
I agree we should improve clarity of the failed case.
The last suggestion requires appending the dialect to all queries which I would like to avoid.
The first one looks good, but I'm not sure writing to the globals considered a good practice, or at least better than overriding the `env`.

Reply:
I think the second option is better here, as it not requiring to switch environments for every dialect. Why don’t you want to add the dialect to the assertions (not the queries themselves)? I think it’s fitting this flow (getting `res` with `cmd`, then comparing it to `expected` with `assertEqual`)

Reply:
The first option I suggested might be an overkill in this case

Reply:
In cluster mode the setting of the default dialect won't get to the shards and thus we need to append the `dialect _` to the queries as well. Otherwise this would have been great.

Reply:
`env.assertEqual(run_command_on_all_shards(env, config_cmd(), 'SET', 'DEFAULT_DIALECT', dialect), ['OK']*env.shardsCount)`

Reply:
Yes but this command is propagated to the shards from the coordinator, and they are not updated with the new dialect, so I think this is not fine unless I'm missing something.
Running commands on all shards is an option, together with adding the dialect to  the assertions.

Reply:
On master (and I think 2.10 as well) `FT.EXPLAIN` and `FT.EXPLAINCLI` are performed locally, as they are not depended on the data, and the schema is known to all the shards